File: src/main/java/org/jukeboxmc/block/BlockMonsterEgg.java
package org.jukeboxmc.block;

import org.jukeboxmc.block.direction.BlockFace;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.math.BlockPosition;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.world.World;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class BlockMonsterEgg extends Block {

    public BlockMonsterEgg() {
        super( "minecraft:monster_egg" );
    }

    @Override
    public void placeBlock( Player player, World world, BlockPosition placePosition, Item itemIndHand, BlockFace blockFace ) {
        this.setMonsterEggStoneType( MonsterEggStoneType.values()[itemIndHand.getMeta()] );
        world.setBlock( placePosition, this );
    }

    @Override
    public Item toItem() {
        return super.toItem().setMeta( this.getMonsterEggStoneType().ordinal() );
    }

    public void setMonsterEggStoneType( MonsterEggStoneType monsterEggStoneType ) {
        this.setState( "monster_egg_stone_type", monsterEggStoneType.name().toLowerCase() );
    }

    public MonsterEggStoneType getMonsterEggStoneType() {
        return this.stateExists( "monster_egg_stone_type" ) ? MonsterEggStoneType.valueOf( this.getStringState( "monster_egg_stone_type" ).toUpperCase()) : MonsterEggStoneType.STONE;
    }

    public enum MonsterEggStoneType {
        STONE,
        COBBLESTONE,
        STONE_BRICK,
        MOSSY_STONE_BRICK,
        CRACKED_STONE_BRICK,
        CHISELED_STONE_BRICK
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockMonsterEgg.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.MonsterEgg
import org.jukeboxmc.api.block.data.MonsterEggStoneType
import org.jukeboxmc.server.block.JukeboxBlock

class BlockMonsterEgg(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates),
    MonsterEgg {

    override fun getMonsterEggStoneType(): MonsterEggStoneType {
        return MonsterEggStoneType.valueOf(this.getStringState("monster_egg_stone_type"))
    }

    override fun setMonsterEggStoneType(value: MonsterEggStoneType): MonsterEgg {
        return this.setState("monster_egg_stone_type", value.name.lowercase())
    }
}